python/hail/typecheck/check.py

```diff
@@ -0,0 +1,206 @@
+
+
+class MultipleTypeChecker(TypeChecker):
+    def __init__(self, checkers):
+        flat_checkers = []
+        for c in checkers:
+            if isinstance(c, MultipleTypeChecker):
+                for cc in c.checkers:
+                    flat_checkers.append(cc)
+            else:
+                flat_checkers.append(c)
+        self.checkers = flat_checkers
+        super(MultipleTypeChecker, self).__init__()
+
+    def check(self, x):
+        return any(c.check(x) for c in self.checkers)
+
+    def expects(self):
+        return '(' + ' or '.join([c.expects() for c in self.checkers]) + ')'
+
+
+class CollectionChecker(TypeChecker):
+    def __init__(self, collection_checker, element_checker):
+        self.cc = collection_checker
+        self.ec = element_checker
+        super(CollectionChecker, self).__init__()
+
+    def check(self, x):
+        passes = self.cc.check(x)
+        return passes and all(self.ec.check(elt) for elt in x)
+
+    def expects(self):
+        return '%s[%s]' % (self.cc.expects(), self.ec.expects())
+
+
+class DictChecker(TypeChecker):
+    def __init__(self, key_checker, value_checker):
+        self.kc = key_checker
+        self.vc = value_checker
+        super(DictChecker, self).__init__()
+
+    def check(self, x):
+        passes = isinstance(x, dict)
+        if passes:
+            for k, v in x.iteritems():
+                passes = passes and self.kc.check(k)
+                passes = passes and self.vc.check(v)
+        return passes
+
+    def expects(self):
+        return 'dict[%s, %s]' % (self.kc.expects(), self.vc.expects())
+
+
+class AnyChecker(TypeChecker):
+    def __init__(self):
+        super(AnyChecker, self).__init__()
+
+    def check(self, x):
+        return True
+
+    def expects(self):
+        return 'any'
+
+
+class LiteralChecker(TypeChecker):
+    def __init__(self, t):
+        self.t = t
+        super(LiteralChecker, self).__init__()
+
+    def check(self, x):
+        return isinstance(x, self.t)
+
+    def expects(self):
+        return extract(self.t)
+
+
+def only(t):
+    if isinstance(t, type) or type(t) is ClassType:
+        return LiteralChecker(t)
+    elif isinstance(t, TypeChecker):
+        return t
+    else:
+        raise RuntimeError("invalid typecheck signature: expected 'type' or 'TypeChecker', found '%s'" % type(t))
+
+
+def oneof(*args):
+    return MultipleTypeChecker([only(x) for x in args])
+
+
+def nullable(t):
+    return oneof(t, NoneType)
+
+
+def listof(t):
+    return CollectionChecker(only(list), only(t))
+
+
+def tupleof(t):
+    return CollectionChecker(only(tuple), only(t))
+
+
+def dictof(k, v):
+    return DictChecker(only(k), only(v))
+
+
+none = only(NoneType)
+
+anytype = AnyChecker()
+
+strlike = oneof(str, unicode)
+
+integral = oneof(int, long)
+
+numeric = oneof(int, long, float)
+
+
+def check_all(f, args, kwargs, checks, is_method):
+    spec = getargspec(f)
+    name = f.__name__
+
+    named_argspec = []
```

Comment:
I'd rename argspec => argname or something.  ArgSpec is an object associated with a function, looks like spec.args is just a list of string, no "spec" anymore.  Makes the set comparison block below a little surprising.